nservicebus/messaging/messages-as-interfaces.md

```diff
@@ -1,21 +1,97 @@
+
+```csharp
+public class UserCreatedFromCampaign : UserCreated
+{
+    public Guid CampaignId { get; set;}
+}
+
+public class RecordStatistics : IHandleMessages<UserCreatedFromCampaign>
+{
+    public async Task Handle(UserCreatedFromCampaign, IMessageHandlerContext context) {
+        // Record statistics
+    }
+}
+```
+
+Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). That is, new functionality has been added without having to change the existing handlers.
+
+In this example, the new event, `UserCreatedFromCampaign`, has a clear relationship with `UserCreated`. However, it's possible that other events could be raised in relation to campaigns, not just users being created. This is where polymorphic routing can help.
+
+## Multiple inheritance to support polymorphic routing
+
+In the updated scenario, there are two concerns: a user is created and a campaign event occurred. The handlers for both events look as follows:
+
+```csharp
+public class SaveUser : IHandleMessages<UserCreated>
+{
+    public async Task Handle(UserCreated message, IMessageHandlerContext) {
+        // Do stuff when the user is created
+    }
+}
+
+public class RecordCampaignActivity : IHandleMessages<CampaignActivityOccurred>
+{
+    public async Task Handle(CampaignActivityOccurred message, IMessageHandlerContext) {
+        // Do stuff when an event related to a campaign happened 
+    }
+}
+```
+
+Now assume the system should perform some action specifically when a user is created from a campaign. For example, perhaps an email is sent to the user thanking them for engaging. This can be accomplished with an event that inherits from both `UserCreated` and `CampaignActivityOccurred` as follows:
+
+```csharp
+public interface UserCreated : IEvent
+{
+    Guid UserId { get; set;}
+}
+
+public interface CampaignActivityOccurred : IEvent
+{
+    Guid CampaignId { get; set;}
+}
+
+public interface UserCreatedFromCampaign : UserCreated, CampaignActivityOccurred
+{
+}
+```
+
+When a user is created from a campaign, the publisher can publish the latter event. It will get handled by the `SaveUser` and `RecordCampaignActivity` handlers defined earlier, and perhaps another handler that explicitly handles `UserCreatedFromCampaign` events. Furthermore, other campaign events can be created that inherit from `CampaignActivityOccurred` without interfering with any functionality around creating users as part of a campaign.
+
+With this approach, events can support multiple inheritance and the system can create messages on the fly that implement these interfaces as described above.
```

Comment:
Something about this sentence doesn't feel right, but I'm not sure if I have anything better. I'm wondering if it's needed?

```suggestion
```